backup: separate tenant spans in include_all_secondary_tenants

msbutler opened the pull request:
Previously they would be merged by distsender which would then cause a single
export response to include multiple tenants, in violation of the sink's
expectations about a single prefix per response.

This patch also enables to remove tenant span adjusting during online restore.

Release note: none.
Epic: none.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/122007)
<!-- Reviewable:end -->

stevendanna commented on pkg/ccl/backupccl/backup_job.go:
```diff
@@ -1521,7 +1521,11 @@ func getTenantInfo(
 		// ever have valid tenant IDs returned to us.
 		prefix := keys.MakeTenantPrefix(roachpb.MustMakeTenantID(tenants[i].ID))
-		spans = append(spans, roachpb.Span{Key: prefix, EndKey: prefix.PrefixEnd()})
+		// We append MaxKey to the tenant's prefix rather than using PrefixEnd since
+		// we want an EndKey that remains within the tenant's prefix and does not
+		// equal the next tenant's start key, as that can cause out spans to get
+		// merged but they need to remain separate for backup prefix elision.
+		spans = append(spans, roachpb.Span{Key: prefix, EndKey: append(prefix, keys.MaxKey...)})
```
We now have this in like 3 places. Perhaps we should move it to a named function somewhere that takes a tenant ID and returns this type of span.

msbutler replied:
yeah, i plan to remove it OR call sites. Keeping this as a draft for now until i fix the interesting failures in CI.

msbutler commented:
RFAL!

msbutler committed:
backup: separate tenant spans in include_all_secondary_tenants

Previously they would be merged by distsender which would then cause a single
export response to include multiple tenants, in violation of the sink's
expectations about a single prefix per response.

This patch also enables to remove tenant span adjusting during online restore.

Release note: none.
Epic: none.

msbutler commented:
@dt you were right. no need for extra logic for checking coverage. Just to make sure, i added a cross version backup tenant check.

msbutler commented:
@dt @stevendanna friendly ping here!

msbutler commented:
TFTR!

bors r=dt
